File: Problems/0295. (H) Find Median from Data Stream/Java/solution.java
class MedianFinder {
  // double heaps

  Queue<Integer> l;
  Queue<Integer> r;

  public MedianFinder() {
    // Java priority queues default to min heaps
    // smaller half, max heap
    this.l = new PriorityQueue(Collections.reverseOrder());
    // larger half, min heap
    this.r = new PriorityQueue();
  }
  
  public void addNum(int num) {
    // both heaps empty
    if (this.l.size() == 0) {
      l.add(num);
      return;
      }
    // if num is smaller than top of smaller half, insert there
    if (num < this.l.peek())
      this.l.add(num);
    else
      this.r.add(num);
    // if size difference is larger than 1, balance
    if (Math.abs(this.l.size() - this.r.size()) > 1) {
      if (this.l.size() > this.r.size()) {
        int temp = this.l.poll();
        this.r.add(temp);
      }
      else {
        int temp = this.r.poll();
        this.l.add(temp);
      }
    }
  }
  
  public double findMedian() {
    // return top of larger heap
    if (this.l.size() > this.r.size())
      return this.l.peek();
    else if (this.r.size() > this.l.size())
      return this.r.peek();
    else
      return (double) (this.l.peek() + this.r.peek()) / 2.0;
  }
}

File: Problems/0295. (H) Find Median from Data Stream/Python3/solution.py
class MedianFinder:
  # double heaps

  def __init__(self):
    # smaller half, max heap
    self._l = []
    # larger half, min heap
    self._r = []
      
  def addNum(self, num: int) -> None:
    # both heaps empty
    if not len(self._l):
      heappush(self._l, -num)
      return
    # if num is smaller than top of smaller half, insert there
    if -num > self._l[0]:
      heappush(self._l, -num)
    else:
      heappush(self._r, num)
    # if size difference of heaps is larger than 1, balance
    if abs(len(self._l) - len(self._r)) > 1:
      if len(self._l) > len(self._r):
        temp = heappop(self._l)
        heappush(self._r, -temp)
      else:
        temp = heappop(self._r)
        heappush(self._l, -temp)

  def findMedian(self) -> float:
    # return top of larger heap
    if len(self._l) > len(self._r):
      return -self._l[0]
    elif len(self._r) > len(self._l):
      return self._r[0]
    # or return average of both tops
    else:
      return (-self._l[0] + self._r[0]) / 2

